dsgrid/dataset/table_format_handler_base.py

```diff
@@ -0,0 +1,96 @@
+import abc
+import logging
+
+from dsgrid.config.project_config import ProjectConfig
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.query.models import AggregationModel
+from dsgrid.query.query_context import QueryContext
+
+
+logger = logging.getLogger(__name__)
+
+
+class TableFormatHandlerBase(abc.ABC):
+    """Base class for table format handers"""
+
+    def __init__(self, project_config: ProjectConfig, dataset_id=None):
+        self._project_config = project_config
+        self._dataset_id = dataset_id
+
+    @abc.abstractmethod
+    def process_aggregations(self, df, aggregations: AggregationModel, context: QueryContext):
+        """Aggregate the dimensional data as specified by aggregations.
+
+        Parameters
+        ----------
+        df : pyspark.sql.DataFrame
+        aggregations : AggregationModel
+        context : QueryContext
+
+        Returns
+        -------
+        pyspark.sql.DataFrame
+
+        """
+
+    @property
+    def project_config(self) -> ProjectConfig:
+        return self._project_config
+
+    @property
+    def dataset_id(self) -> str:
+        return self._dataset_id
+
+    @abc.abstractmethod
+    def add_columns(self, df, dimension_query_names, context: QueryContext, aggregation_allowed):
+        """Add columns to the dataframe.
+
+        Parameters
+        ----------
+        df : pyspark.sql.DataFrame
+        dimension_query_names : list
+        context : QueryContext
+        aggregation_allowed : bool
+            Set to False if adding a column is not allowed to change the load values in df.
+
+        """
+
+    def convert_columns_to_query_names(self, df):
+        # All columns start off as base dimension names but need to be query names.
+        base_to_query_name_mapping = self.project_config.get_base_dimension_to_query_name_mapping()
+        columns = set(df.columns)
+        for dim_type in DimensionType:
+            if dim_type == DimensionType.TIME:
+                time_dim = self._project_config.get_base_dimension(dim_type)
+                time_cols = time_dim.get_timestamp_load_data_columns()
+                # TODO DT: Should we enforce that projects can only have one time column?
```

Comment:
we can't right because of Tempo? Unless we always explode time cols to timestamps

Reply:
The problem here is that for all stacked dimensions there is one column per dimension, and so we can rename the column to be the query name. Is this true for time? @elainethale @mooneyme Do you have an opinion?

Reply:
I think we decided to leave as-is for now and revisit with TEMPO time PR